File: PART3/ch02/ch02_01/java-kotlin/src/main/java/org/example/ex06/Exam06.java
package org.example.ex06;

interface Bark {
    void bark();
}

abstract class Animal implements Bark {

    private final String name;
    private int age;


    public int getAge() {
        return age;
    }

    public void setAge(int age) {
        this.age = age;
    }

    public Animal(String name){
        this.name = name;
    }

    public void eat(){
        System.out.println(name+" 식사 시작 합니다.");
    }

    public String getName() {
        return name;
    }
}

class Dog extends Animal{

    public Dog(){
        super("");
    }

    public Dog(String name){
        super(name);
    }

    @Override
    public void bark() {
        System.out.println(this.getName()+" : 멍멍");
    }
}


public class Exam06 {

    public Exam06(){
        var dog = new Dog("퍼피");
        dog.setAge(10);
        var age = dog.getAge();
        dog.eat();
        dog.bark();
    }

    public static void main(String[] args){
        new Exam06();
    }
}

File: PART3/ch01/kotlin-example/kotlin-example/src/main/kotlin/ex06/Exam06.kt
package ex06

fun main(){
    val dog = Dog("해피")

    println(dog.age)

    dog.age = 20

    println(dog.age)

    dog.eat()
    dog.bark()
}

interface Bark{
    fun bark()
}

interface Temp{
    fun hi()
}

abstract class Animal(
    private val name: String?="",
): Bark {
    // body
    fun eat(){
        println("$name 식사시작 합니다.")
    }
}

class Dog(
    private val name:String?=null,
) : Animal(name), Temp {

    var age: Int?=0

    override fun bark() {
        println("$name : 멍멍")
    }

    override fun hi() {
        println("hi")
    }

}